File: scripts/repro-inkcrash.py
#!/usr/bin/env python3
# DIRIGE o harness `repro-inkcrash.mjs` num PTY REAL e escreve a sequência CSI-u de tecla
# FUNCIONAL (`\x1b[57414u`, kitty kbd proto) no stdin do filho — o caminho EXATO do crash.
# O harness escreve marcadores num ARQUIVO de log (REPRO_LOG) p/ não disputar com o repaint
# do Ink no stdout. Lemos esse log: BOOT / ALIVE n / INKCRASH / EXIT code ⇒ VIVO ou MORTO.
#
# Uso: GUARD=off python3 scripts/repro-inkcrash.py   (baseline: deve crashar)
#      python3 scripts/repro-inkcrash.py             (fix:      deve sobreviver)
import os, pty, select, time, sys, re, tempfile

HARNESS = os.path.join(os.path.dirname(__file__), "repro-inkcrash.mjs")
SEQ = b"\x1b[57414u"  # CSI 57414 u — kitty functional key (o byte que crasha o Ink)
NODE = os.environ.get("NODE_BIN", "node")


def run():
    logf = tempfile.NamedTemporaryFile(prefix="repro-inkcrash-", suffix=".log", delete=False)
    logpath = logf.name
    logf.close()
    open(logpath, "w").close()  # zera

    pid, fd = pty.fork()
    if pid == 0:  # filho: TTY real no stdin/stdout
        os.environ["REPRO_LOG"] = logpath
        os.execvp(NODE, [NODE, HARNESS])
        os._exit(127)

    def read_log():
        try:
            with open(logpath) as f:
                return f.read()
        except OSError:
            return ""

    def drain_pty(t):
        end = time.time() + t
        while time.time() < end:
            r, _, _ = select.select([fd], [], [], 0.05)
            if r:
                try:
                    os.read(fd, 65536)
                except OSError:
                    return

    # 1) espera o boot (marcador BOOT no log).
    booted = False
    end = time.time() + 3.0
    while time.time() < end:
        drain_pty(0.1)
        if "BOOT" in read_log():
            booted = True
            break
    print(f"[boot]    montou (marcador BOOT): {booted}")

    log_before = read_log()
    ticks_before = len(re.findall(r"ALIVE \d+", log_before))

    # 2) ENVIA a sequência perigosa pelo PTY real (vai pro stdin do Ink).
    os.write(fd, SEQ)
    print(f"[send]    sequencia escrita no PTY: {SEQ!r}")

    # 3) observa ~2s: VIVO (heartbeats novos) ou MORTO (crash)?
    drain_pty(2.0)
    log_after = read_log()

    crashed = "INKCRASH" in log_after
    ticks_after_total = len(re.findall(r"ALIVE \d+", log_after))
    new_ticks = ticks_after_total - ticks_before

    # 4) colhe o exit-code do filho.
    try:
        os.close(fd)
    except OSError:
        pass
    code = None
    try:
        _, status = os.waitpid(pid, 0)
        code = os.waitstatus_to_exitcode(status)
    except OSError:
        pass

    m = re.search(r"INKCRASH (.*)", log_after)
    if m:
        print(f"[crash]   mensagem: {m.group(1).strip()}")
    print(f"[observe] heartbeats ALIVE NOVOS apos a sequencia: {new_ticks}")
    print(f"[observe] INKCRASH no log: {crashed}")
    print(f"[observe] exit-code do filho: {code}")
    print(f"[observe] log final: {read_log().strip().splitlines()[-4:]!r}")

    try:
        os.unlink(logpath)
    except OSError:
        pass

    survived = (not crashed) and new_ticks >= 1
    print()
    print("RESULTADO:", "SOBREVIVEU (fix OK)" if survived else "CRASHOU/MORREU (baseline)")
    return 0 if survived else 1


if __name__ == "__main__":
    sys.exit(run())

File: scripts/repro-inkcrash.mjs
// REPRO/PROVA task #18 — CRASH do Ink em sequência CSI-u de tecla FUNCIONAL (kitty kbd
// proto). Renderiza o App REAL com o Ink numa TTY (dirigido por um PTY real via
// `repro-inkcrash.py`, que escreve `\x1b[57414u` no stdin do filho). Sem o fix (GUARD=off),
// o Ink crasha em `use-input.js:73` (`input.startsWith` sobre undefined). Com o fix (filtro
// raw, default), a sequência é descartada ANTES do Ink e o app SOBREVIVE.
//
// Os MARCADORES de protocolo vão num ARQUIVO de log (REPRO_LOG) — NÃO no stdout/stderr,
// que o Ink repinta/limpa. O driver lê esse arquivo p/ decidir VIVO/MORTO:
//   · `BOOT`        — montou e está vivo (composer pronto).
//   · `ALIVE t`     — heartbeat: ainda VIVO no tick t.
//   · `INKCRASH …`  — uncaughtException (o crash do Ink).
//   · `EXIT code`   — saída (limpa = 0).
// Sem backend (o model é um stub). Bug é puramente de UI/parse.

import { appendFileSync } from 'node:fs';
import React from 'react';
import { render } from 'ink';
import { ThemeProvider } from '../packages/cli/dist/ui/theme/context.js';
import { resolveTheme } from '../packages/cli/dist/ui/theme/theme.js';
import { App } from '../packages/cli/dist/session/App.js';
import { SessionController } from '../packages/cli/dist/session/controller.js';
import { TuiAskResolver } from '../packages/cli/dist/ask/ask-resolver.js';
import { PolicyPermissionEngine } from '@hiperplano/aluy-cli-core';
import { installCsiUGuard } from '../packages/cli/dist/session/csi-u-guard.js';

const LOG = process.env.REPRO_LOG ?? '/tmp/repro-inkcrash.log';
const log = (line) => {
  try {
    appendFileSync(LOG, line + '\n');
  } catch {
    /* best-effort */
  }
};

const env = { LANG: 'en_US.UTF-8', TERM: 'xterm-256color' };

// O guard SÓ é instalado quando NÃO for o baseline (GUARD=off reproduz o crash pré-fix;
// default = fix ligado, espelhando o que o run.tsx faz antes do render).
if (process.env.GUARD !== 'off') installCsiUGuard(process.stdin);

const ports = {
  fs: {
    async readFile() {
      return '';
    },
    async writeFile() {},
    async exists() {
      return false;
    },
  },
  shell: {
    async exec() {
      return { stdout: '', stderr: '', exitCode: 0 };
    },
  },
  search: {
    async search() {
      return [];
    },
  },
};

const controller = new SessionController({
  model: {
    async call() {
      return { request_id: 'r', content: '', finish_reason: 'stop' };
    },
  },
  permission: new PolicyPermissionEngine(),
  ports,
  askResolver: new TuiAskResolver(),
  meta: { cwd: '/proj/aluy-vau', tier: 'aluy-flux', tokens: 1280, windowPct: 12 },
  flush: { intervalMs: 0 },
});
controller.dismissBoot();

// O crash do Ink vira uncaughtException — registra e sai com código DISTINTO do limpo.
process.on('uncaughtException', (e) => {
  log(`INKCRASH ${e && e.message ? e.message : e}`);
  log('EXIT 42');
  process.exit(42);
});

const theme = resolveTheme({ env });
const { waitUntilExit } = render(
  React.createElement(
    ThemeProvider,
    { theme },
    React.createElement(App, { controller, animate: false, bootMs: 0 }),
  ),
  { exitOnCtrlC: false },
);

log('BOOT');
let tick = 0;
const hb = setInterval(() => {
  tick += 1;
  log(`ALIVE ${tick}`);
  if (tick >= 6) {
    clearInterval(hb);
    log('EXIT 0');
    process.exit(0); // sobreviveu a tudo: saída limpa.
  }
}, 200);

await waitUntilExit().catch(() => {});
